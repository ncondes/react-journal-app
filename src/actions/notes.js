import Swal from 'sweetalert2';
import { db } from "../firebase/firebase-config";
import { collection, addDoc, doc, updateDoc, deleteDoc } from "firebase/firestore"; 
import { types } from '../types/types';
import { loadNotes } from '../helpers/loadNotes';
import { fileUpload } from '../helpers/fileUpload';


export const startNewNote = () => {
    return async ( dispatch, getState ) => {

        const { uid } = getState().auth;
        try {
            const docRef = await addDoc( collection( db, `${uid}/journal/notes`), {
                title: '',
                body: '',
                date: new Date().getTime(),
            });
            dispatch( activeNote( docRef.id, {
                title: '',
                body: '',
                date: new Date().getTime(),
            }))
            dispatch( addNewNote( docRef.id, {
                title: '',
                body: '',
                date: new Date().getTime(),
            }))
        } catch (error) {
            console.log('Error adding document: ', error );
        }
    }
}


export const activeNote = ( id, note ) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note,
    }
});


export const addNewNote = ( id, note ) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note,
    }
})


export const startLoadingNotes = ( uid ) => {
    return async ( dispatch ) => {
        const notes = await loadNotes( uid );
        dispatch( setNotes( notes ) );
    } 
}


export const setNotes = ( notes ) => ({
    type: types.notesLoad,
    payload: notes,
});


export const startSaveNote = ( note ) => {
    return async ( dispatch, getState ) => {

        const { uid } = getState().auth;
        if ( !note.url ) {
            delete note.url
        }
        const noteToFirestore = { ...note };
        delete noteToFirestore.id;
        const noteRef = doc( db, `${ uid }/journal/notes/${ note.id }`)

        await updateDoc( noteRef, noteToFirestore )
        dispatch( refreshNote( note.id, noteToFirestore ) );
        Swal.fire('Saved', note.title, 'success')
    }
}


export const refreshNote = ( id, note ) => ({
    type: types.notesUpdated,
    payload: {
        id,
        note: {
            id,
            ...note
        }
    }
})
 

export const startUploading = ( file ) => {
    return async ( dispatch, getState ) => {
        const { active: activeNote } = getState().notes;
        Swal.fire({
            title: 'Uploading',
            text: 'Please wait.',
            allowOutsideClick: false,
            didOpen: () => {
                Swal.showLoading();
            }
        });
        const fileUrl = await fileUpload( file );
        activeNote.url = fileUrl;
        dispatch( startSaveNote( activeNote ) )
        Swal.close();
    }
}


export const startDeleting = ( id ) => {

    return async ( dispatch, getState ) => {
        const { uid } = getState().auth;

        Swal.fire({
            title: 'Are you sure?',
            text: "You won't be able to revert this!",
            icon: 'warning',
            showCancelButton: true,
            confirmButtonColor: '#3085d6',
            cancelButtonColor: '#d33',
            confirmButtonText: 'Yes, delete it!'
        }).then( async (result) => {
            if (result.isConfirmed) {
                await deleteDoc( doc( db, `${ uid }/journal/notes/${ id }` ) )
                dispatch( deleteNote( id ) );
                Swal.fire(
                    'Deleted!',
                    'Your file has been deleted.',
                    'success'
                )
            }
        })
    }
}


export const deleteNote = ( id ) => ({
    type: types.notesDelete,
    payload: id,
})


export const noteLogout = () => ({
    type: types.notesLogoutCleaning,
})