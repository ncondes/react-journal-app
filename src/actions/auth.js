import { types } from '../types/types';
import { getAuth, signInWithPopup, FacebookAuthProvider, createUserWithEmailAndPassword, updateProfile, signInWithEmailAndPassword, signOut } from "firebase/auth";
import { googleAuthProvider } from '../firebase/firebase-config';
import { finishLoading, startLoading } from './ui';
import Swal from 'sweetalert2';
import { noteLogout } from './notes';

const auth = getAuth();

export const startLoginEmailPassword = ( email, password ) => {
    return ( dispatch ) => {

        dispatch( startLoading() )

        signInWithEmailAndPassword( auth, email, password )
            .then( ({ user }) => {

                dispatch(
                    login( user.uid, user.displayName )
                )

                dispatch( finishLoading() )

            })
            .catch( error => {
                Swal.fire({
                    title: 'Error!',
                    text: error.message,
                    icon: 'error',
                    confirmButtonText: 'OK',
                })
                dispatch( finishLoading() )
            })
    }
}

export const startRegisterWithEmailPasswordName = ( name, email, password ) => {

    return ( dispatch ) => {

        createUserWithEmailAndPassword( auth, email, password )
            .then( async ({ user }) => {
                await updateProfile( auth.currentUser, {
                    displayName: name,
                })
                dispatch(
                    login( user.uid, user.displayName )
                )
            })
            .catch ( error => {
                console.log( error );
            }) 
    }
}

export const startGoogleLogin = () => {
    return ( dispatch ) => {

        signInWithPopup( auth, googleAuthProvider )
            .then( (result) => {
                const user = result.user
                dispatch(
                    login( user.uid, user.displayName )
                )
            })

    }
}

export const startFacebookLogin = () => {
    return ( dispatch ) => {
        const provider = new FacebookAuthProvider();
        signInWithPopup( auth, provider )
            .then( ({ user }) => {
                dispatch(
                    login( user.uid, user.displayName )
                )
            })
    }
}

export const login = ( uid, displayName ) => ({
        type: types.login,
        payload: {
            uid,
            displayName,
        },
})

export const startLogout = () => {
    return async ( dispatch ) => {
        await signOut( auth )
            .then( () => {
                dispatch( logout() )
                dispatch( noteLogout() )
            })
            .catch( error => {
                console.log( error );
            })
    }
} 

export const logout = () => ({
    type: types.logout
})