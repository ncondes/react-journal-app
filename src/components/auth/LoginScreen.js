import { useDispatch, useSelector } from 'react-redux';
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth';


export const LoginScreen = () => {


    const dispatch = useDispatch();
    const { ui } = useSelector(state => state)

    const [ formValues, handleInputChange ] = useForm({
        email: 'Fernanda@gmail.com',
        password: '123456'
    })

    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        dispatch( startLoginEmailPassword( email, password ) );
    }

    const handleGoogleLogin = () => {
        dispatch( startGoogleLogin() );
    }

    return (
        <>
        	<h2 className='auth__title animate__animated animate__fadeIn animate__fast'>Login</h2>
            <form
                className='auth__form animate__animated animate__fadeIn animate__fast'
                onSubmit={ handleLogin }
            >
                <label className='auth__label'>
                    Email
                    <div className='auth__input-container'>
                        <img
                            src='../assets/icons/mail-icon.png'
                            alt='mail icon'
                        />
                        <input
                            type='email'
                            placeholder='Enter your Email'
                            name='email'
                            autoComplete='off'
                            value={ email }
                            onChange={ handleInputChange }
                        />
                    </div>
                </label>
                <label className='auth__label'>
                    Password
                    <div className='auth__input-container'>
                        <img
                            src='../assets/icons/key-icon.png'
                            alt='key icon'
                        />
                        <input
                            type='password'
                            placeholder='************'
                            name='password'
                            value={ password }
                            onChange={ handleInputChange }
                        />
                    </div>
                </label>
                <button
                    type='submit'
                    className='btn btn-white btn-block mt-5 mb-5'
                    disabled={ ui.loading }
                >
                    Login
                </button>
                <div className='auth__social'>
                    <p>- or -</p>
                    <p>Login with social</p>
                    <div className='auth__social-links'>
                        <div className='btn-white btn-circle'>
                            <img
                                src='../assets/icons/facebook-icon.png'
                                alt='facebook icon'
                            />
                        </div>
                        <div
                            className='btn-white btn-circle'
                            onClick={ handleGoogleLogin }
                        >
                            <img
                                src='../assets/icons/google-icon.png'
                                alt='google icon'
                            />
                        </div>
                    </div>
                </div>
                <p className='auth__bottom'>
                    Don't have and account?
                    <Link to='/auth/register'> Sign up</Link>
                </p>

            </form>
        </>
    )
}
