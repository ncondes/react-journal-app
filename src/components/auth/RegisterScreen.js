import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import validator from 'validator';
import { useDispatch } from 'react-redux';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';
import Swal from 'sweetalert2';


export const RegisterScreen = () => {

    const dispatch = useDispatch();

    // import useSelector from react-redux this to select one of the states of the redux
    // const { errorMessage } = useSelector( state => state.ui ); 

    const [ formValues, handleInputChange ] = useForm({
        name: 'Fernanda',
        email: 'Fernanda@gmail.com',
        password: '123456',
        confirmPassword: '123456',
    })

    const {
        name,
        email,
        password,
        confirmPassword
    } = formValues;

    const handleRegister = ( e ) => {
        e.preventDefault();
        if ( isFormValid() ) {
            dispatch( startRegisterWithEmailPasswordName( name, email, password ) )
        }
    }

    const isFormValid = () => {

        if ( name.trim().length === 0 ) {
            Swal.fire({
                title: 'Error!',
                text: 'Full Name is required',
                icon: 'error',
                confirmButtonText: 'OK',
            })
            dispatch( setError( 'Full Name is required' ) )
            return false
        } else if ( !validator.isEmail( email ) ) {
            Swal.fire({
                title: 'Error!',
                text: 'Email is not valid',
                icon: 'error',
                confirmButtonText: 'OK',
            })
            dispatch( setError( 'Email is not valid' ) )
            return false
        } else if ( password !== confirmPassword || password.length < 5 ) {
            Swal.fire({
                title: 'Error!',
                text: 'Password should be at least 6 cahracters and match each other',
                icon: 'error',
                confirmButtonText: 'OK',
            })
            dispatch( setError( 'Password should be at least 6 cahracters and match each other' ) )
            return false
        }

        dispatch( removeError() )
        return true;
    }

    return (
        <>
            <h2 className='auth__title animate__animated animate__fadeIn animate__fast'>Sign Up</h2>
            <form
                className='auth__form animate__animated animate__fadeIn animate__fast'
                onSubmit={ handleRegister }
            >

                <label className='auth__label'>
                    Full Name
                    <div className='auth__input-container'>
                        <img
                            src='../assets/icons/user-icon.png'
                            alt='mail icon'
                        />
                        <input
                            type='text'
                            placeholder='Enter your Full Name'
                            name='name'
                            autoComplete='off'
                            value={ name }
                            onChange={ handleInputChange }
                        />
                    </div>
                </label>
                <label className='auth__label'>
                    Email
                    <div className='auth__input-container'>
                        <img
                            src='../assets/icons/mail-icon.png'
                            alt='mail icon'
                        />
                        <input
                            type='email'
                            placeholder='Enter your Email'
                            name='email'
                            autoComplete='off'
                            value={ email }
                            onChange={ handleInputChange }
                        />
                    </div>
                </label>
                <label className='auth__label'>
                    Password
                    <div className='auth__input-container'>
                        <img
                            src='../assets/icons/key-icon.png'
                            alt='key icon'
                        />
                        <input
                            type='password'
                            placeholder='************'
                            name='password'
                            value={ password }
                            onChange={ handleInputChange }
                        />
                    </div>
                </label>
                <label className='auth__label'>
                    Confirm Password
                    <div className='auth__input-container'>
                        <img
                            src='../assets/icons/key-icon.png'
                            alt='key icon'
                        />
                        <input
                            type='password'
                            placeholder='************'
                            name='confirmPassword'
                            value={ confirmPassword }
                            onChange={ handleInputChange }
                        />
                    </div>
                </label>
                <button
                    type='submit'
                    className='btn btn-white btn-block mt-5 mb-5'
                >
                    Register
                </button>

                <p className='auth__bottom'>
                    Have an account?
                    <Link to='/auth/login'> Login</Link>
                </p>

            </form>
        </>
    )
}
