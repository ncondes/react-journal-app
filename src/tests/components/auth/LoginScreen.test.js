import { mount } from 'enzyme';
import { LoginScreen } from '../../../components/auth/LoginScreen';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import '@testing-library/jest-dom';
import { MemoryRouter } from 'react-router-dom'
import { startGoogleLogin, startLoginEmailPassword } from '../../../actions/auth';


jest.mock('../../../actions/auth', () => ({
    startGoogleLogin: jest.fn(),
    startLoginEmailPassword: jest.fn(),
}))



const middlewares = [ thunk ];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {},
    ui: {
        loading: false,
        errorMessage: null,
    }
};

let store = mockStore( initState );
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={ store }>
        <MemoryRouter>
            <LoginScreen />
        </MemoryRouter>
    </Provider>
);

describe('Pruebas en el Nombre del componente', () => {


    beforeEach(() => {
        store = mockStore( initState );
        jest.clearAllMocks();
    })

    test('Debe de mostrarse correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    });

    test('Debe de disparar la accion de startGoogleLogin', () => {
        
        wrapper.find('.btn-google').prop('onClick')();
        expect( startGoogleLogin ).toHaveBeenCalled();

    });

    test('Debe de disparar el startLogin con los respectivos argumentos', () => {
        
        wrapper.find('.auth__form').prop('onSubmit')({
            preventDefault(){}
        });
        expect( startLoginEmailPassword ).toHaveBeenCalledWith( '', '' );

    })
    

})