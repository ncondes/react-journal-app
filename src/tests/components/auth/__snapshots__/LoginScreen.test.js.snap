// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el Nombre del componente Debe de mostrarse correctamente 1`] = `
Array [
  <h2
    className="auth__title animate__animated animate__fadeIn animate__fast"
  >
    Login
  </h2>,
  <form
    className="auth__form animate__animated animate__fadeIn animate__fast"
    onSubmit={[Function]}
  >
    <label
      className="auth__label"
    >
      Email
      <div
        className="auth__input-container"
      >
        <img
          alt="mail icon"
          src="../assets/icons/mail-icon.png"
        />
        <input
          autoComplete="off"
          name="email"
          onChange={[Function]}
          placeholder="Enter your Email"
          type="email"
          value=""
        />
      </div>
    </label>
    <label
      className="auth__label"
    >
      Password
      <div
        className="auth__input-container"
      >
        <img
          alt="key icon"
          src="../assets/icons/key-icon.png"
        />
        <input
          name="password"
          onChange={[Function]}
          placeholder="************"
          type="password"
          value=""
        />
      </div>
    </label>
    <button
      className="btn btn-white btn-block mt-5 mb-5"
      disabled={false}
      type="submit"
    >
      Login
    </button>
    <div
      className="auth__social"
    >
      <p>
        - or -
      </p>
      <p>
        Login with social
      </p>
      <div
        className="auth__social-links"
      >
        <div
          className="btn-white btn-circle"
          onClick={[Function]}
        >
          <img
            alt="facebook icon"
            src="../assets/icons/facebook-icon.png"
          />
        </div>
        <div
          className="btn-white btn-circle btn-google"
          onClick={[Function]}
        >
          <img
            alt="google icon"
            src="../assets/icons/google-icon.png"
          />
        </div>
      </div>
    </div>
    <p
      className="auth__bottom"
    >
      Don't have and account?
      <a
        href="/auth/register"
        onClick={[Function]}
      >
         Sign up
      </a>
    </p>
  </form>,
]
`;
